File: MATLAB/gradient_descent.m
% gradient_descent.m
% Gradient Descent Algorithm for Multivariable Function Minimization in MATLAB

function [minimum, iterations, converged] = gradient_descent(func, grad_func, initial_guess, learning_rate, max_iterations, tolerance)
    % Inputs:
    %   func - Handle to the multivariable function to minimize
    %   grad_func - Handle to the gradient function of func
    %   initial_guess - Initial starting point for the algorithm
    %   learning_rate - Step size for gradient descent
    %   max_iterations - Maximum number of iterations allowed
    %   tolerance - Convergence threshold for stopping criterion
    
    % Initialize variables
    x = initial_guess;
    converged = false;
    
    for i = 1:max_iterations
        grad = grad_func(x);
        x_new = x - learning_rate * grad;
        
        % Check for convergence
        if norm(x_new - x) < tolerance
            converged = true;
            break;
        end
        
        x = x_new;
    end
    
    % Output results
    minimum = x;
    iterations = i;
end

File: Python/gradient_descent.py
# gradient_descent.py
# Gradient Descent Algorithm for Multivariable Function Minimization in Python

import numpy as np

class GradientDescent:
    def __init__(self, func, gradient, learning_rate=0.01, max_iterations=1000, tolerance=1e-6):
        self.func = func
        self.gradient = gradient
        self.learning_rate = learning_rate
        self.max_iterations = max_iterations
        self.tolerance = tolerance
    
    def find_minimum(self, initial_guess):
        x = np.array(initial_guess, dtype=float)
        
        for i in range(self.max_iterations):
            grad = np.array(self.gradient(x))
            x_new = x - self.learning_rate * grad
            
            # Check for convergence
            if np.linalg.norm(x_new - x) < self.tolerance:
                return {"minimum": x_new, "iterations": i+1, "converged": True}
            
            x = x_new
        
        return {"minimum": x, "iterations": self.max_iterations, "converged": False}
